### ISSUE TITLE
Speeding up cachetxt additions

### ISSUE THREAD

**Issue by mattsinc** [ROCmSoftwarePlatform/MIOpenGEMM#36]
I was wondering if any of you have any feedback on ways I could speed up the cachetxt additions to the kernel cache?  I know these files change occasionally, but they are relatively constant, so it seems like creating a .so file (with the values for the different components set in an array or something similar) and loading from that would improve performance of this step significantly.  However, since this hasn't been done already, I wanted to check if my understanding about this component was off and there's a reason why the different kernels are being loaded into MIOpenGEMM every time it executes?

Thanks,
Matt

**Comment by newling** [comment]
Having one / several source files for different cachetxt files sounds like a good idea to me. Last time I checked (a few months ag0) the compilation wasn't all that slow though (?)

Note that there are no OpenCL kernels (no kernel source, no binary, nothing) when MIOpenGEMM is compiled, and so a large .cachetxt just means a large map / unordered_map. I'm not sure how large this needs to be to feel the slowdown. Also, actually generating a new cachetxt (which involves generating and compiling many OpenCL kernels) is surely slower than compiling MIOpenGEMM? So maybe I'm missing something here.

**Comment by newling** [comment]
I don't understand why it is so slow. It looks like there are O(10^3) entries in the cachetxt files. If it's taking O(1) seconds to (i) build the std::map and (ii) find the nearest match for a given geometry, that's like O(10^6) cycles per entry which seems like too many (each entry is just a few strings).

I think this pull request is related to the slowness : https://github.com/ROCmSoftwarePlatform/MIOpenGEMM/pull/30
Maybe @zjing14 has some ideas.

Matt are you proposing to have the std::map constructed at compile time? I'm not sure how to do this. 

James

**Comment by zjing14** [comment]
In #30, we had to increase the threshold of cache search time, since we added many entries into cachetxt. Otherwise, the MIOpenGEMM failed in Jenkins.

**Comment by mattsinc** [comment]
@newling, yeah it appears that it takes much longer now (thanks @zjing14).  I don't have a perfect plan for how to fix it, but perhaps a std::map that is just set up by default in the code would work?  Then if/when we wanted to add a new entry, instead of adding it to a text file, we could just add another entry to the map.

Matt
